packages/@lwc/template-compiler/src/shared/estree.ts

```diff
@@ -18,6 +18,18 @@ export function isArrayExpression(node: t.BaseNode): node is t.ArrayExpression {
     return node.type === 'ArrayExpression';
 }
 
+export function isObjectExpression(node: t.BaseNode): node is t.ObjectExpression {
+    return node.type === 'ObjectExpression';
+}
+
+export function isExpression(node: t.BaseNode): node is t.Expression {
+    return node.type === 'Expression';
```

Comment:
This will not work, `t.Expression` is a union type. https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/estree/index.d.ts#L222-L229

```ts
// prettier-ignore
const EXPRESSION_TYPES = new Set([
    'ThisExpression', 'ArrayExpression', 'ObjectExpression', 'FunctionExpression',
    'ArrowFunctionExpression', 'YieldExpression', 'Literal', 'UnaryExpression', 'UpdateExpression',
    'BinaryExpression', 'AssignmentExpression', 'LogicalExpression', 'MemberExpression',
    'ConditionalExpression', 'CallExpression', 'NewExpression', 'SequenceExpression', 
    'TemplateLiteral', 'TaggedTemplateExpression', 'ClassExpression', 'MetaProperty', 'Identifier', 
    'AwaitExpression', 'ImportExpression', 'ChainExpression' 
]);

export function isExpression(node: t.BaseNode): node is t.Expression {
    return EXPRESSION_TYPES.has(node.type);
}
```

Reply:
Yeah, I think I made a mistake in my array-checking logic. Removed the `t.isExpression` check entirely, and now it looks like deep arrays are correctly being hoisted: 636abcad4